### PR TITLE
Resilver restarts unnecessarily when it encounters errors

### PR DESCRIPTION
**Pull request by jwpoduska**
### Motivation and Context
We have observed that if any errors are encountered during resilver, the entire resilver is restarted from the beginning every time it finishes, never completing, as opposed to finishing the resilver and possibly restarting using DTL_SCRUB.

### Description
When a resilver finishes, vdev_dtl_reassess is called to hopefully
excise DTL_MISSING (amongst other things). If there are errors during
the resilver, they are tracked in DTL_SCRUB, as spelled out in the
block comment in vdev.c. DTL_SCRUB is in-core only, so it can only
be used if the pool was online for the whole resilver. This state is
tracked with the spa_scrub_started flag, which only gets set when
the scan is initialized. Unfortunately, this flag gets cleared right
before vdev_dtl_reassess gets called, so if there are any errors
during the scan, DTL_MISSING will never get excised and the resilver
will just continually restart. This fix simply moves clearing that
flag until after the call to vdev_dtl_reasses.

In addition, if a pool is imported and already has scn_errors > 0,
this change will restart the resilver immediately instead of doing
the rest of the scan and then restarting it from the beginning. On
the other hand, if scn_errors == 0 at import, then no errors have
been encountered so far, so the spa_scrub_started flag can be safely
set.

### How Has This Been Tested?
Using zinject, I confirmed that the resilver restarted after finishing without the fix and completed with the fix. In a similar way, I confirmed that if no errors were encountered before exporting/importing the pool and then injecting an error before resilver finished that resilver would complete and use DTL_SCRUB. Lastly, I confirmed by injecting errors before export/import that the resilver restarted after import instead of going to the end and then restarting completely.

A new ZTS test was added to verify that resilver will not restart after finishing if the required DTL's are available.

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

File: module/zfs/dsl_scan.c [hunk 1]
```diff
@@ -542,6 +542,22 @@ dsl_scan_init(dsl_pool_t *dp, uint64_t txg)
...
+			 * If a resilver is in progress and there are already
+			 * errors, restart it instead of finishing this scan and
+			 * then restarting it. If there haven't been any errors
+			 * then remember that the incore DTL is valid.
+			 */
+			if (scn->scn_phys.scn_errors > 0) {
+				scn->scn_restart_txg = txg;
+				zfs_dbgmsg("resilver can't excise DTL_MISSING "
+				    "when finished; restarting in txg %llu",
+				    (longlong_t)scn->scn_restart_txg);
```
Comment (behlendorf) [inline module/zfs/dsl_scan.c:556]: typo: longlong_t -> u_longlong_t

Comment (jwpoduska) [reply module/zfs/dsl_scan.c]: I made that change, though, there are a lot of instances of this issue in dsl_scan.c. I also added a tests to verify the fix.

**Comment by jwpoduska** [comment]
Looks like one of the bsd runs of the new test didn't like 'stat --format=%i' to get the object id, so I changed it to use 'ls -i'.

**Comment by scineram** [comment]
This should be in 0.8 IMO.

**Comment by agilob** [comment]
Just found this PR in release notes o 0.8.5

>We have observed that if any errors are encountered during resilver, the entire resilver is restarted from the beginning every time it finishes, never completing, as opposed to finishing the resilver and possibly restarting using DTL_SCRUB.

@jwpoduska do you think this will fix issue that's been haunting me? https://github.com/openzfs/zfs/issues/11018

**Pull request merged**